File: code/templates/R/01_cohort_identification_template.R
# 01 — Cohort identification (worked example)
#
# This is the one fully worked R template in this project. It demonstrates the CLIF idiom end to
# end; steps 02–04 are skeletons for you to fill in.
#
# Run from the PROJECT ROOT, e.g.:
#     Rscript code/01_cohort_identification_template.R
#
# What it does:
#   1. Load config + the tables this project needs (errors clearly if one is missing)
#   2. Build a simple cohort  (EDIT the criteria for your project)
#   3. Save patient-level working data to output/intermediate_phi/  (NEVER shared, git-ignored)
#   4. Save a shareable aggregate to output/final_no_phi/           (no row-level data, every cell n >= 10)

library(here)
library(tidyverse)
library(arrow)

# 1. Configuration — reads config/config.json (see config/README.md).
#    config.json defaults to the bundled clif_demo/ dataset, so this runs out of the box;
#    point data_directory at your CLIF tables for a real run.
source("utils/config.R")
data_directory <- config$data_directory
filetype       <- config$filetype

# Helper: read a CLIF table by name, with a clear error if the file is missing.
read_clif_table <- function(name) {
  path <- file.path(data_directory, paste0("clif_", name, ".", filetype))
  if (!file.exists(path)) stop("Missing required CLIF table: ", path)
  if (filetype == "parquet") {
    collect(open_dataset(path))
  } else if (filetype == "csv") {
    read_csv(path, show_col_types = FALSE)
  } else {
    stop("Unsupported filetype: ", filetype, " (use 'csv' or 'parquet')")
  }
}

hospitalization <- read_clif_table("hospitalization")
adt             <- read_clif_table("adt")

# 2. Build the cohort — EDIT these criteria for your project.
start_date <- as.Date("2020-01-01")
end_date   <- as.Date("2021-12-31")

inpatient_ids <- adt %>%
  filter(location_category %in% c("Ward", "ICU")) %>%
  distinct(hospitalization_id) %>%
  pull(hospitalization_id)

cohort <- hospitalization %>%
  filter(as.Date(admission_dttm) >= start_date,
         as.Date(admission_dttm) <= end_date,
         age_at_admission >= 18,
         hospitalization_id %in% inpatient_ids)

message("Cohort: ", n_distinct(cohort$hospitalization_id), " hospitalizations")

# 3. Save patient-level working data (git-ignored, never leaves your site).
dir.create(here("output", "intermediate_phi"), showWarnings = FALSE, recursive = TRUE)
write_parquet(cohort, here("output", "intermediate_phi", "cohort.parquet"))

# 4. Save a shareable aggregate — never report cells with n < 10 (re-identification risk).
dir.create(here("output", "final_no_phi"), showWarnings = FALSE, recursive = TRUE)
cohort_summary <- cohort %>%
  mutate(age_group = cut(age_at_admission, breaks = c(18, 40, 65, Inf),
                         right = FALSE, labels = c("18-39", "40-64", "65+"))) %>%
  group_by(age_group) %>%
  summarise(n_hospitalizations = n_distinct(hospitalization_id), .groups = "drop") %>%
  filter(n_hospitalizations >= 10)   # data-security rule
write_csv(cohort_summary, here("output", "final_no_phi", "cohort_summary.csv"))
message("Wrote ", nrow(cohort_summary), " aggregate row(s) to output/final_no_phi/")

File: code/templates/Python/01_cohort_identification_template.py
"""
01 — Cohort identification (worked example)

This is the one fully worked Python template in this project. It demonstrates the CLIF idiom end to
end; steps 02–04 are skeletons for you to fill in.

Run from the PROJECT ROOT, e.g.:
    uv run python code/01_cohort_identification_template.py

What it does:
  1. Initialize clifpy from config/config.json
  2. Load the tables this project needs
  3. Build a simple cohort  (EDIT the criteria for your project)
  4. Save patient-level working data to output/intermediate_phi/  (NEVER shared, git-ignored)
  5. Save a shareable aggregate to output/final_no_phi/           (no row-level data, every cell n >= 10)
"""

from pathlib import Path

import pandas as pd
from clifpy import ClifOrchestrator

INTERMEDIATE_DIR = Path("output/intermediate_phi")  # patient-level — stays at your site
FINAL_DIR = Path("output/final_no_phi")             # aggregate — safe to share

# 1. Initialize clifpy from the config file (it reads data_directory / filetype / timezone).
#    config.json defaults to the bundled clif_demo/ dataset, so this runs out of the box;
#    point data_directory at your CLIF tables for a real run.
co = ClifOrchestrator(config_path="config/config.json")

# 2. Load the tables this project needs. clifpy raises a clear error if a table is missing.
co.initialize(tables=["patient", "hospitalization", "adt"])
hospitalization = co.hospitalization.df.copy()
adt = co.adt.df.copy()

# 3. Build the cohort — EDIT these criteria for your project.
START_DATE, END_DATE = "2020-01-01", "2021-12-31"
hospitalization["admission_dttm"] = pd.to_datetime(hospitalization["admission_dttm"])

# CLIF category values are lowercase.
inpatient_ids = adt.loc[
    adt["location_category"].isin(["ward", "icu"]), "hospitalization_id"
].unique()

cohort = hospitalization[
    hospitalization["admission_dttm"].between(START_DATE, END_DATE)
    & (hospitalization["age_at_admission"] >= 18)
    & hospitalization["hospitalization_id"].isin(inpatient_ids)
]
print(f"Cohort: {cohort['hospitalization_id'].nunique()} hospitalizations")

# 4. Save patient-level working data (git-ignored, never leaves your site).
INTERMEDIATE_DIR.mkdir(parents=True, exist_ok=True)
cohort.to_parquet(INTERMEDIATE_DIR / "cohort.parquet", index=False)

# 5. Save a shareable aggregate — never report cells with n < 10 (re-identification risk).
FINAL_DIR.mkdir(parents=True, exist_ok=True)
cohort = cohort.assign(
    age_group=pd.cut(
        cohort["age_at_admission"],
        bins=[18, 40, 65, 120],
        right=False,
        labels=["18-39", "40-64", "65+"],
    )
)
summary = (
    cohort.groupby("age_group", observed=True)["hospitalization_id"]
    .nunique()
    .reset_index(name="n_hospitalizations")
)
summary = summary[summary["n_hospitalizations"] >= 10]  # data-security rule
summary.to_csv(FINAL_DIR / "cohort_summary.csv", index=False)
print(f"Wrote {len(summary)} aggregate row(s) to {FINAL_DIR}")
